### ISSUE TITLE
Update SV loading pipeline

### ISSUE THREAD

**Issue by lynnpais** [broadinstitute/seqr#3064]
As the calls are undergoing changes, we need to update our SV loading pipeline to meet the most recent specs.

**Comment by hanars** [comment]
A test subset can be found here: [CMG-GREGoR > Structural Variation folder](https://drive.google.com/drive/u/0/folders/1-hC5mrMms2nNnUDqXGJQw6qNwh0G2Fil)

**Comment by hanars** [comment]
also should check if there are any new columns we were not previously loading to determine if we should add them

**Comment by ShifaSZ** [comment]
A summary of the changes of the new genome structural variant data format:

- The ‘ALT’ and ‘FORMAT’ columns have no changes. The CPX_TYPEs have no changes.

- The some ‘FILTER’ values are removed which are LOW_CALL_RATE, UNSTABLE_AF_PCRMINUS, FAIL_FINAL_RECALIBRATION, FAIL_BATCH_EFFECT, FAIL_OUTLIER_REMOVAL, and FAIL_minGQ

- There are many changes in the INFO columns.
   - No population information, such as AN, AC, AF, gnomAD SVID, and gnomAD AF
   - Prediction columns have been changed completely. The new prediction columns include: PREDICTED_LOF, PREDICTED_COPY_GAIN, PREDICTED_DUP_PARTIAL, PREDICTED_INTRONIC, PREDICTED_INV_SPAN, PREDICTED_UTR, PREDICTED_NONCODING_SPAN, PREDICTED_NONCODING_BREAKPOINT, PREDICTED_NEAREST_TSS, PREDICTED_INTERGENIC, PREDICTED_PROMOTER, PREDICTED_BREAKEND_EXONIC, PREDICTED_INTRAGENIC_EXON_DUP, PREDICTED_MSV_EXON_OVERLAP, PREDICTED_TSS_DUP, PREDICTED_PARTIAL_EXON_DUP

**Comment by karynne7** [comment]
There is gnomad freq information, but it's annotated like the following:
```##INFO=<ID=gnomad_v2.1_sv_SVID,Number=1,Type=String,Description="SVID of an overlapping event in gnomad used for external allele frequency annotation.">
##INFO=<ID=gnomad_v2.1_sv_AF,Number=1,Type=Float,Description="Allele frequency (for biallelic sites) or copy-state frequency (for multiallelic sites) of an overlapping event in gnomad.">
##INFO=<ID=gnomad_v2.1_sv_AFR_AF,Number=1,Type=Float,Description="Allele frequency (for biallelic sites) or copy-state frequency (for multiallelic sites) of an overlapping event in gnomad.">
##INFO=<ID=gnomad_v2.1_sv_AMR_AF,Number=1,Type=Float,Description="Allele frequency (for biallelic sites) or copy-state frequency (for multiallelic sites) of an overlapping event in gnomad.">
##INFO=<ID=gnomad_v2.1_sv_EAS_AF,Number=1,Type=Float,Description="Allele frequency (for biallelic sites) or copy-state frequency (for multiallelic sites) of an overlapping event in gnomad.">
##INFO=<ID=gnomad_v2.1_sv_EUR_AF,Number=1,Type=Float,Description="Allele frequency (for biallelic sites) or copy-state frequency (for multiallelic sites) of an overlapping event in gnomad.">
```

**Comment by karynne7** [comment]
I have been using bcftools to fill tags for AN,AC,AF:
```
bcftools +fill-tags tmp.vcf.bgz -Ob -o filled_tag.vcf -- -t AN,AC,AF
```
And this script to go between PREDICTED* columns and formatting:
```
#!/usr/bin/env bash

vcf=$1
out=$2

sed -e 's/PREDICTED_/PROTEIN_CODING__/g' $vcf | \
	sed -e 's/PROTEIN_CODING__MSV_EXON_OVERLAP/PROTEIN_CODING__MSV_EXON_OVR/g' | \
	sed -e 's/gnomad_v2.1_sv/gnomAD_V2/g' | \
	sed -e 's/PROTEIN_CODING__NONCODING/NONCODING/g' | \
	awk -vOFS="\t" '{sub(/.*\./,"",$3);print $0}' > $out
```

**Comment by ShifaSZ** [comment]
@karynne7 I found I missed a whole session of the headers. I see the frequency columns now.

**Comment by hanars** [comment]
> Thanks, @karynne7. I found I missed a whole session of the headers. I see the frequency columns now.

Can you give an updated summary of what changed then?

**Comment by ShifaSZ** [comment]
@hanars, I have just updated my comment above.

**Comment by hanars** [comment]
> The prediction columns have been changed.

Is there a 1-to-1 mapping from the old to new columns, or are there predictions we no longer have/ new predictions we did not used to have

> New frequency columns
 
Just to clarify, are all the old frequency columns still there? 
Also, are `gnomAD_V2_AC_AF` and `gnomAD_V2_AN_AF` just the AC and AN for gnomad? Or are they actually an AF (the difference being AC/AN are integers and AF is a fraction, AF = AC/AN)

**Comment by ShifaSZ** [comment]
> Is there a 1-to-1 mapping from the old to new columns, or are there predictions we no longer have/ new predictions we did not used to have

Most of the 'new' predictions can be mapped to the 'old' ones. But some of them are completely new. I will update the mapping in the above comment.

**Comment by ShifaSZ** [comment]
> Just to clarify, are all the old frequency columns still there?
> Also, are `gnomAD_V2_AC_AF` and `gnomAD_V2_AN_AF` just the AC and AN for gnomad? Or are they actually an AF (the difference being AC/AN are integers and AF is a fraction, AF = AC/AN)

All the old frequency columns are still there, including `AN`, `AC`, `AF`, `gnomAD_V2_SVID`, and `gnomAD_V2_AF`. The type of `gnomAD_V2_AC_AF` and `gnomAD_V2_AN_AF` is `float`. I'm not sure what they are exactly for.

**Comment by hanars** [comment]
We will want to update the SV consequences we allow users to filter on in seqr to reflect these new consequences, but that can be done as a separate ticket from this. I'll follow up with Alba about the one prediction that seems to be missing in the new data, and about the new gnomad scores

**Comment by hanars** [comment]
> Some ‘FILTER’ values are removed

Do we use these in out pipeline at all? I only see us using the `filters` field

**Comment by ShifaSZ** [comment]
The value of the `filters` field in the MatrixTable row is from the `FILTER` column in the VCF file. We include the FILTER values in the loading pipeline, excluding `PASS` and `BOTHSIDES_SUPPORT`.

**Comment by hanars** [comment]
Alba confirmed that `gnomAD_V2_AC_AF` and `gnomAD_V2_AN_AF` are in fact the AC and AN for that population, so we will want to add those, and the when we actually take this live we can show them

**Comment by hanars** [comment]
Alba confirmed that `gnomAD_V2_AC_AF` and `gnomAD_V2_AN_AF` are in fact the AC and AN for that population, so we will want to add those, and the when we actually take this live we can show them

**Comment by hanars** [comment]
Alba says that the "missing" filters were added in post processing, so we won't expect those going forward and I think we don't need to worry about that change

**Comment by ShifaSZ** [comment]
> Alba confirmed that `gnomAD_V2_AC_AF` and `gnomAD_V2_AN_AF` are in fact the AC and AN for that population, so we will want to add those, and the when we actually take this live we can show them

I will add the `gnomad_svs_AC` and `gnomad_svs_AN` fields to the output Elasticsearch index. See the [document](https://docs.google.com/document/d/1_FyERQB0vxcUo1yKm0C_fgQCl8sFuIHfInNYgoOTy08/edit#) for details.

**Comment by ShifaSZ** [comment]
> We have a couple options here: we can update the pipeline to map both `PREDICTED_INTRAGENIC_EXON_DUP` and `PREDICTED_PARTIAL_EXON_DUP` to `DUP_LOF`, or we can load those both separately and update the seqr search UI to include those as filter options and replace the current `DUP_LOF`

How about we load both `PREDICTED_INTRAGENIC_EXON_DUP` and `PREDICTED_PARTIAL_EXON_DUP` separately?

**Comment by ShifaSZ** [comment]
The new prediction fields `PREDICTED_NONCODING_SPAN` and `PREDICTED_NONCODING_BREAKPOINT` are not genes. Do we still add them to the `sortedTranscriptConsequences` list? If yes, what are the values we should use for `gene_symbol`
`gene_id` fields?

**Comment by hanars** [comment]
Can you please make another ticket for updating seqr to support the updated SV data? We will need to make sure that live before we can use this updated pipeline

**Comment by hanars** [comment]
> The new prediction fields PREDICTED_NONCODING_SPAN and PREDICTED_NONCODING_BREAKPOINT are not genes

We should omit these, same as we do nearest TSS

**Comment by ShifaSZ** [comment]
> Can you please make another ticket for updating seqr to support the updated SV data? We will need to make sure that live before we can use this updated pipeline

Sure, I will. BTW. Does the seqr need to support both SV data formats?

**Comment by hanars** [comment]
> Sure, I will. BTW. Does the seqr need to support both SV data formats?

Yes

**Comment by ShifaSZ** [comment]
PRs:
https://github.com/broadinstitute/seqr-loading-pipelines/pull/380
https://github.com/broadinstitute/seqr-loading-pipelines/pull/381

**Comment by ShifaSZ** [comment]
Another change in the data format:
- The ‘EV’ field in the ‘FORMAT’ column has been changed from a single value to an array of values.

**Comment by hanars** [comment]
> Another change in the data format:
> 
> * The ‘EV’ field in the ‘FORMAT’ column has been changed from a single value to an array of values.

Are we using that field anywhere?

**Comment by ShifaSZ** [comment]
We don't use the field. But it makes our test fixture data incompatible.

**Comment by hanars** [comment]
There are inputs to the tests that are faked lines from a VCF. Those will change. There are outputs that we actually test to make sure they are correctly parsed. Since the output does not include this field, the fact that you update the fake input lines should have no impact on the tested output line

**Comment by hanars** [comment]
Completed in https://github.com/broadinstitute/seqr-loading-pipelines/pull/381
